File: TwoPointer/8.string-to-integer-atoi.go
package leetcode

import (
	"math"
	"unicode"
)

// @leet start
func myAtoi(s string) int {
	var (
		num, i int
		sign   = 1
	)
	for i < len(s) && s[i] == ' ' {
		i++
	}
	s = s[i:]
	if len(s) == 0 {
		return 0
	}
	switch s[0] {
	case '+':
		s = s[1:]
	case '-':
		sign = -1
		s = s[1:]
	}

	for _, r := range s {
		if !unicode.IsDigit(r) {
			break
		}
		num = num*10 + int(r-'0')
		if sign*num <= math.MinInt32 {
			return math.MinInt32
		}
		if sign*num >= math.MaxInt32 {
			return math.MaxInt32
		}
	}

	return num * sign
}

// @leet end

File: TwoPointer/8.string-to-integer-atoi.py
# @leet start
class Solution:
    def myAtoi(self, s: str) -> int:
        i, n = 0, len(s)
        sign = 1
        while i < n and s[i] == " ":
            i += 1
        if i < n and s[i] == "+" or s[i] == "-":
            if s[i] == "-":
                sign = -1
            i += 1
        num = 0
        while i < n and s[i].isdigit():
            num = num * 10 + ord(s[i]) - ord("0")
            i += 1
        num *= sign
        minR = -(2**31)
        maxR = 2**31 - 1

        if num > maxR:
            return maxR
        elif num < minR:
            return minR
        else:
            return num


# @leet end

